exporter/otlpexporter/exporter.go

```diff
@@ -0,0 +1,274 @@
+)
+
+// Crete new exporter and start it. The exporter will begin connecting but
+// this function may return before the connection is established.
+func newExporter(config *Config) (*exporterImp, error) {
+	e := &exporterImp{}
+	e.config = config
+	e.disconnectedCh = make(chan bool, 1)
+	e.stopCh = make(chan bool)
+	e.backgroundConnectionDoneCh = make(chan bool)
+
+	var err error
+	e.dialOpts, err = configgrpc.GrpcSettingsToDialOptions(e.config.GRPCSettings)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := e.start(); err != nil {
+		return nil, err
+	}
+	return e, nil
+}
+
+// Start dials to the collector, establishing a connection to it. It also
+// initiates the Config and Trace services by sending over the initial
+// messages that consist of the node identifier. Start invokes a background
+// connector that will reattempt connections to the collector periodically
+// if the connection dies.
+func (e *exporterImp) start() error {
+	var err = errAlreadyStarted
+	e.startOnce.Do(func() {
+		e.mutex.Lock()
+		stopped := e.stopped
+		e.started = true
+		e.mutex.Unlock()
+
+		if stopped {
+			err = errStopped
+			return
+		}
+
+		// An optimistic first connection attempt to ensure that
+		// applications under heavy load can immediately process
+		// data. See https://github.com/census-ecosystem/opencensus-go-exporter-ocagent/pull/63
+		if err := e.connect(); err == nil {
+			e.setStateConnected()
+		} else {
+			e.setStateDisconnected(err)
+		}
+		go e.indefiniteBackgroundConnection()
+
+		err = nil
+	})
+
+	return err
+}
+
+func (e *exporterImp) stop() error {
+	e.stopOnce.Do(func() {
+		e.mutex.Lock()
+		e.stopped = true
+		started := e.started
+		e.mutex.Unlock()
+
+		if !started {
+			// Not yet started, nothing to do.
+			return
+		}
+
+		// Started. Signal to stop.
+		close(e.stopCh)
+
+		// Wait until stopped.
+		<-e.backgroundConnectionDoneCh
+	})
+	return nil
+}
+
+// Send a trace or metrics request to the server. "perform" function is expected to make
+// the actual gRPC unary call that sends the request. This function implements the
+// common OTLP logic around request handling such as retries and throttling.
+func (e *exporterImp) exportRequest(ctx context.Context, perform func(ctx context.Context) error) error {
+
+	expBackoff := backoff.NewExponentialBackOff()
+
+	// Spend max 15 mins on this operation. This is just a reasonable number that
+	// gives plenty of time for typical quick transient errors to resolve.
+	expBackoff.MaxElapsedTime = time.Minute * 15
+
+	for {
+		// Send to server.
+		err := perform(ctx)
+
+		if err == nil {
+			// Request is successful, we are done.
+			return nil
+		}
+
+		// We have an error, check gRPC status code.
+
+		status := status.Convert(err)
+
+		statusCode := status.Code()
+		if statusCode == codes.OK {
+			// Not really an error, still success.
+			return nil
+		}
+
+		// Now, this is this a real error.
+
+		if !shouldRetry(statusCode) {
```

Comment:
TODO: we don't want this logic on the exporter since it breaks the composability of components.